parlai/agents/style_gen/modules.py

```diff
@@ -0,0 +1,266 @@
+                  (Only used if not self.personality_as_label)
+        :return:
+            - the model's predicted per-class scores.
+              (FloatTensor[bsz, len(class_list)])
+        """
+
+        if self.personality_as_label:
+            # All tokens go into the encoder and classification is learned from that.
+            assert len(xs) == 1
+            # Only one input allowed
+            bsz = xs[0].size(0)
+            encoder_states = self.encoder(*xs)
+            inputs = self.START.detach().expand(bsz, 1)
+            # Generate most likely class given start token as input
+            latent, _ = self.decoder(inputs, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.squeeze(dim=1))
+        else:
+            # Tokens are split between the encoder and decoder and classification is
+            # learned from both.
+            text_vec, label_vec = xs
+            encoder_states = self.encoder(text_vec)
+            latent, _ = self.decoder(label_vec, encoder_states)
+            # latent: [bsz, seqlen, emb_dim]
+            scores = self.classifier_head(latent.mean(dim=1))
+
+        return scores
+
+
+class BatchWithPersonalities(AttrDict):
+    """
+    Adds a 'personalities' field to the batch in the case where personality information
+    is not encoded in any other field.
+    """
+
+    def __init__(self, personalities=None, **kwargs):
+        super().__init__(personalities=personalities, **kwargs)
+
+
+class TransformerDecoderWithEmbeds(TransformerDecoder):
+    def forward(self, input, encoder_state, embedded_input=None, incr_state=None):
+        """
+        Forward pass with the ability to pass in token-embedded inputs.
+        """
+        # TODO: perhaps reduce the amount of code duplicated from TransformerDecoder.
+        #  This would require modularizing several snippets of code inside
+        #  TransformerDecoder methods.
+
+        encoder_output, encoder_mask = encoder_state
+
+        if input is not None:
+            seq_len = input.size(1)
+            positions = input.new(seq_len).long()
+        else:
+            seq_len = embedded_input.size(1)
+            positions = embedded_input.new(seq_len).long()
+        positions = torch.arange(seq_len, out=positions).unsqueeze(0)
+
+        if incr_state is not None:
+            # We're doing incremental decoding, so select only the most recent position
+            if input is not None:
+                input = input[:, -1:]
+            if embedded_input is not None:
+                embedded_input = embedded_input[:, -1:, :]
+            if positions is not None:
+                positions = positions[:, -1:]
+        else:
+            incr_state = {}
+
+        if embedded_input is not None:
+            tensor = embedded_input  # No need to copy because we only reassign below
+        else:
+            tensor = self.embeddings(input)
```

Comment:
Hmm, I can't replace this with a call to `TransformerEncoder.forward_embedding()` unless that function allows token embeddings to be passed in instead of token IDs - see my note on the following comment